File: 380-insert-delete-getrandom-o1/insert-delete-getrandom-o1.py
class RandomizedSet:

    def __init__(self):
        self.num_map = {}
        self.list_num = []

    def insert(self, val: int) -> bool:
        res = val not in self.num_map
        if res:
            self.num_map[val]=len(self.list_num)
            self.list_num.append(val)
        return res

    def remove(self, val: int) -> bool:
        res = val in self.num_map
        if res:
            indx = self.num_map[val]
            last_val = self.list_num[-1]
            self.list_num[indx] = last_val
            self.num_map[last_val] = indx
            self.list_num.pop()
            del self.num_map[val]
        return res

    def getRandom(self) -> int:
        return random.choice(self.list_num)


# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()

File: 380-insert-delete-getrandom-o1/insert-delete-getrandom-o1.java
import java.util.*;

class RandomizedSet {
    private Map<Integer, Integer> numMap;
    private List<Integer> listNum;
    private Random random;
    
    public RandomizedSet() {
        numMap = new HashMap<>();
        listNum = new ArrayList<>();
        random = new Random();
    }
    
    public boolean insert(int val) {
        boolean res = !numMap.containsKey(val);
        if (res) {
            numMap.put(val, listNum.size());
            listNum.add(val);
        }
        return res;
    }
    
    public boolean remove(int val) {
        boolean res = numMap.containsKey(val);
        if (res) {
            int idx = numMap.get(val);
            int lastVal = listNum.get(listNum.size() - 1);
            listNum.set(idx, lastVal);
            numMap.put(lastVal, idx);
            listNum.remove(listNum.size() - 1);
            numMap.remove(val);
        }
        return res;
    }
    
    public int getRandom() {
        return listNum.get(random.nextInt(listNum.size()));
    }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet obj = new RandomizedSet();
 * boolean param_1 = obj.insert(val);
 * boolean param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */

